src/zarr/abc/store.py

```diff
@@ -81,6 +81,13 @@ async def empty(self) -> bool: ...
     @abstractmethod
     async def clear(self) -> None: ...
 
+    @abstractmethod
+    def with_mode(self, mode: AccessModeLiteral) -> Self:
+        """
+        Return a new store pointing to the same location with a new mode.
```

Comment:
This should document the semantics:

1. same backing data
2. require that `_open` be called? We currently just call it from `make_store_path`, which immediately calls `ensure_open`.